native/dsterf.go

```diff
@@ -0,0 +1,283 @@
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/lapack"
+)
+
+// Dsterf computes all eigenvalues of a symmetric tridiagonal matrix using the
+// Pal-Walker-Kahan variant of the QL or QR algorithm.
+//
+// d contains the diagonal elements of the tridiagonal on entry, and contains the
+// eigenvalues in ascending order on exit. d must have length at least n, or
+// Dsterf will panic.
+//
+// e contains the off-digonal elements of the tridiagonal on entry, and is
+// overwritten during the call to Dsterf.
+func (impl Implementation) Dsterf(n int, d, e []float64) (ok bool) {
+	if n < 0 {
+		panic(nLT0)
+	}
+	if n == 0 {
+		return true
+	}
+	if len(d) < n {
+		panic(badD)
+	}
+	if len(e) < n-1 {
+		panic(badE)
+	}
+
+	// Determine the unit roundoff for this environment.
+	eps := dlamchE
+	eps2 := eps * eps
+	safmin := dlamchS
+	safmax := 1 / safmin
+	ssfmax := math.Sqrt(safmax) / 3
+	ssfmin := math.Sqrt(safmin) / eps2
+
+	// Compute the eigenvalues of the tridiagonal matrix.
+	maxit := 30
+	nmaxit := n * maxit
+	jtot := 0
+
+	l1 := 0
+
+	var anorm, c, gamma, r, rte, s, sigma float64
+	var iscale, l, lend, lendsv, lsv, m int
+
+Ten:
+	if l1 > n-1 {
+		goto OneSeventy
+	}
+	if l1 > 0 {
+		e[l1-1] = 0
+	}
+	for m = l1; m < n-1; m++ {
+		if math.Abs(e[m]) <= math.Sqrt(math.Abs(d[m]))*math.Sqrt(math.Abs(d[m+1]))*eps {
+			e[m] = 0
+			goto Thirty
+		}
+	}
+	m = n - 1
+
+Thirty:
+	l = l1
+	lsv = l
+	lend = m
+	lendsv = lend
+	l1 = m + 1
+	if lend == 0 {
+		goto Ten
+	}
+
+	// Scale submatrix in rows and columns l to lend.
+	anorm = impl.Dlanst(lapack.MaxAbs, lend-l+1, d[l:], e[l:])
+	iscale = 0
+	if anorm == 0 {
+		goto Ten
+	}
+	if anorm > ssfmax {
+		iscale = 1
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmax, lend-l, 1, e[l:], n)
+	} else if anorm < ssfmin {
+		iscale = 2
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l+1, 1, d[l:], n)
+		impl.Dlascl(lapack.General, 0, 0, anorm, ssfmin, lend-l, 1, e[l:], n)
+	}
+
+	for i := l; i < lend; i++ {
+		e[i] *= e[i]
+	}
+
+	// Choose between QL and QR iteration.
+	if math.Abs(d[lend]) < math.Abs(d[l]) {
+		lend = lsv
+		l = lendsv
+	}
+	if lend >= l {
+		// QL Iteration.
+		// Look for small subdiagonal element.
+	Fifty:
+		if l != lend {
+			for m = l; m < lend; m++ {
+				if math.Abs(e[m]) <= eps2*(math.Abs(d[m]*d[m+1])) {
+					goto Seventy
+				}
+			}
+		}
+		m = lend
+	Seventy:
+		if m < lend {
+			e[m] = 0
+		}
+		p := d[l]
+		if m == l {
+			goto Ninety
+		}
+		// If remaining matrix is 2 by 2, use Dlae2 to compute its eigenvalues.
+		if m == l+1 {
+			rte = math.Sqrt(e[l])
+			rt1, rt2 := impl.Dlae2(d[l], rte, d[l+1])
```

Comment:
Do this assignment directly to `d[l]` and `d[l+1]`, `rt1` and `rt2` are not otherwise used.